### ISSUE TITLE
Smartphone gamepad app for Remote RetroPad

### ISSUE THREAD

**Issue by duduke** [libretro/RetroArch#5379]
I'm creating this Issue to serve as a bounty for someone to create an Android and iOS app that will use Remote RetroPad to enable using your smartphone as a remote gamepad for RetroArch.

This is especially useful for a mobile LAKKA install where you do not need to carry and gamepad with you.

**Comment by RobLoach** [comment]
The two I've seen are:

- [nJoy](https://play.google.com/store/apps/details?id=nemex.nJoy&feature=search_result)
- [BT Controller](https://play.google.com/store/apps/details?id=softigloo.btcontroller)

**Comment by duduke** [comment]
Those act as gamepads, I'm looking for something that can talk to RetroPad remote directly. Also, those are only for Android, I'm looking for iOS support as well.

**Comment by fr500** [comment]
There is a core already (I made it) but @twinaphex disabled it somehow and I haven't found out how to reenable it.

It should work already once he re-enables it.

**Comment by duduke** [comment]
I’ve used that core a while ago when you created it. Can it be adapted to a standalone app for both iOS and Android?

**Comment by fr500** [comment]
hmmm I guess but it makes no sense making a standalone app out of it.
Just make a new one, it uses UDP it should be trivial to implement.

@twinaphex can you enable the builtin net retropad core again? 
Not really sure why it was disabled in the first place.

**Comment by duduke** [comment]
I’ve had a go at creating an iOS app that would talk to RetroArch via UDP - unfortunately, I lack any kind of coding background, so didn’t really get anywhere.

**Comment by fr500** [comment]
why does it have to be a standalone app?

**Comment by duduke** [comment]
Doesn’t have to be. but, unfortunately you cannot have RetroArch on iOS without jailbreak - which kinda limit the whole “anyone can use their smartphone as gamepad” idea.

**Comment by twinaphex** [comment]
Yeah, this could be an opportunity for our Apple Developer account which so far doesn't have a single app on it.

**Comment by jmichael2497** [comment]
First off, thanks for your contributions to RetroArch / LibRetro and sharing them with the world, next...

Hope this suggestion helps simplify rather than add more complexity.

I would really like a lightweight system to use my phone as a gamepad versus the current method of netplay sync which is a bit awkward to get matching cores and entering ip addresses, and falling out of sync quagmire.

There are pros and cons to making yet another task specific specialized device app for just a couple of eco-systems versus a possibly more inclusive and open solution since you already have some useful code:

Gamepad support could be built into RetroArch itself similar to how Netplay is built in.

Mobile devices could just run the browser based RetroArch set to load the gamepad core by default.

Perhaps served out by the RetroArch host, since players would connect to it over local wifi anyway.

NOTE: I've been meaning to sign up, just to comment on this, but as I haven't found documentation or been able to use the gamepad core mentioned other than to talk to itself from client to host... I'm not sure how much that core would help as a ready to go part in this idea.

**Comment by fr500** [comment]
I already implemented this years ago but someone broke it.

**Comment by twinaphex** [comment]
I don't know about what I could have done to break it. I would highly encourage you to see if you can get it back to work again. I don't believe I touch any netplay parts, @GregorR does most of the networking code.

**Comment by GregorR** [comment]
I do most of the networking code insofar as netplay is most of the networking code and I do netplay, but I've never touched remote retropad. It's not inconceivable that it has some interdependency with netplay, and if it does, I've definitely broken it, but I'd say it's unlikely.

**Comment by twinaphex** [comment]
@GregorR Would you maybe like to look into why it doesn't work, regardless of who broke it? It'd be great if we could get it to work again.

As for me, I never could get it to work personally.

**Comment by ICHx** [comment]
Hey devs, I have one question for retropad remote.

To setup the ip on android do I need to tap 192+168+1 times to enter 192.168.0.1?  Is there any shortcut? :sweat_smile:

**Comment by daviddavo** [comment]
@ICHx I've opened an Issue at #6873

**Comment by ICHx** [comment]
I am currently using https://github.com/RetroPie/RetroPie-Setup/wiki/Virtual-Gamepad as a rememdy for my gamepad needs.

**Comment by jmichael2497** [comment]
> I am currently using https://github.com/RetroPie/RetroPie-Setup/wiki/Virtual-Gamepad

that is kinda like what i meant (and as i realize my earlier post was maybe a bit unclear).

i'm basically picturing something like those popular multiplayer drop in with your mobile party game systems.

(like by a certain long lived non-tv quiz game show that lets you "screw your neighbor", and now they have other stuff like drawing games, etc, but trying not to commercial name drop in open source environment.)

they probably use webrtc or similar for connection by anyone with a device that has a relatively modern browser.

all that is needed by the end users is to just load a specific url (ideally assisted with qrcode presented on the monitor to avoid manual typing) and then they have a web based gamepad.

this should be served from the game system host device over at least local shared wifi network, but ideally as a direct hotspot for best performance.

hope that helped clarify.

**Comment by duduke** [comment]
I think this project can be used to provide drop-in multiplayer through smartphone controllers. (and it's open source).
[https://snex.io/]
